### ISSUE TITLE
SDPA with nested backend: expose a way to avoid recomputing data layout information

### ISSUE THREAD

**Issue by fxmarty** [pytorch/pytorch#108770]
### 🚀 The feature, motivation and pitch

Hi, `torch.nn.functional.scaled_dot_product_attention` has a very nice torch.nested backend that allows to use e.g. flash attention with variable length inputs. In general, this is useful when doing training (fine-tuning) with padding, or during batched inference.

However, profiling SDPA and looking at the code, it happens that some overhead exists for the torch.nested backend, notably https://github.com/pytorch/pytorch/blob/49e964cad6c6f0241e08adb619bec637763888a8/aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp#L228 which computes cumulative sequence length and maximum sequence length, especially this [to](https://github.com/pytorch/pytorch/blob/49e964cad6c6f0241e08adb619bec637763888a8/aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp#L257C14-L257C14) operation. However, those are usually constants over different layers.

![image](https://github.com/pytorch/pytorch/assets/9808326/eed6e1bf-3611-4ebe-b102-82bac1fdbb4a)
(with code borrowed from the nice path in https://github.com/pytorch/pytorch/issues/105913, here max_seqlen=4096, pad percentage=0.4, batch_size=16, head_dim=128, num_heads=32)

As computing those stats only once [is recommended](https://github.com/Dao-AILab/flash-attention/issues/11#issuecomment-1156681278) for flash attention, it would be neat to be able to indeed compute them only once using PyTorch interface.

Related: https://github.com/huggingface/transformers/pull/25598

### Alternatives

/

### Additional context

/
